src/control/server/config/server.go

```diff
@@ -541,56 +631,34 @@ func (cfg *Server) Validate(log logging.Logger, hugepageSize int) (err error) {
 				sysXSCount++
 			}
 		}
-
 		log.Debug(msg)
 	}
 
-	if cfg.NrHugepages < 0 || cfg.NrHugepages > math.MaxInt32 {
-		return FaultConfigNrHugepagesOutOfRange(cfg.NrHugepages, math.MaxInt32)
-	}
-
-	// Calculate hugepages based on total target count if bdevs in config.
-	if cfgTargetCount > 0 {
-		if cfg.DisableHugepages {
-			return FaultConfigHugepagesDisabled
-		}
-
-		// Calculate minimum number of hugepages for all configured engines.
-		minHugePages, err := common.CalcMinHugePages(hugePageSize, cfgTargetCount+sysXSCount)
-		if err != nil {
+	if len(cfg.Engines) > 1 {
+		if err := cfg.validateMultiEngineConfig(log); err != nil {
 			return err
 		}
+	}
 
-		// If the config doesn't specify hugepages, use the minimum. Otherwise, validate
-		// that the configured amount is sufficient.
-		if cfg.NrHugepages == 0 {
-			var msgSysXS string
-			if sysXSCount > 0 {
-				msgSysXS = fmt.Sprintf(" and %d sys-xstreams", sysXSCount)
-			}
-			log.Debugf("calculated nr_hugepages: %d for %d targets%s", minHugePages,
-				cfgTargetCount, msgSysXS)
-			cfg.NrHugepages = minHugePages
-		}
+	if cfg.NrHugepages < 0 || cfg.NrHugepages > math.MaxInt32 {
+		return FaultConfigNrHugepagesOutOfRange(cfg.NrHugepages, math.MaxInt32)
+	}
 
-		if cfg.NrHugepages < minHugePages {
-			return FaultConfigInsufficientHugePages(minHugePages, cfg.NrHugepages)
-		}
+	if err := hugepageSetCount(log, cfgTargetCount, sysXSCount, mi.HugepageSizeKiB, cfg); err != nil {
```

Comment:
I think we really need to stop putting these kinds of side-effects in the `Validate()` methods. Just from the standpoint of being about to reason about a system, `Validate()` should be a read-only operation that either succeeds with no error or returns an error indicating why the configuration is not valid. It shouldn't be doing stuff to make the configuration valid. That's confusing and harder to test, IMO.

Reply:
I have moved the `SetNrHugepages` and `SetRamdisk` functions out of validate in the second batch of code review related changes.